### ISSUE TITLE
Requesting guidance for generator based object factories

### ISSUE THREAD

**Issue by dnnsthnnr** [mwiatrzyk/pydio#7]
Hi there 👋 

I absolutely love the simplicity of the library!

I have a question around the functionality explained here in the docs: https://pydio.readthedocs.io/en/latest/quickstart.html#using-generator-based-object-factories

My goal is to achieve the following setup with pydio. If an error is raised during the usage of the context manager for the sqlalchemy connection with transaction context the transaction is rolled back, otherwise committed.
```
from contextlib import contextmanager

from sqlalchemy import create_engine, Connection

from pydio.injector import Injector
from pydio.provider import Provider


class ITodoItemStorage:
    pass


class SQLTodoStorage(ITodoItemStorage):

    def __init__(self, connection: Connection):
        self._conn = connection

    def create(self):
        raise ValueError()


@contextmanager
def get_connection(engine):
    with engine.begin() as c:
        try:
            yield c
        except:
            print('rollback')
            c.rollback()
        else:
            print('commit')
            c.commit()


class Application:

    def __init__(self):
        self._engine = create_engine("postgresql://postgres:postrgres@localhost:5432/postgres")

    def create(self):
        with get_connection(self._engine) as c:
            SQLTodoStorage(c).create()

if __name__ == '__main__':
    app = Application()
    app.create()
```

According to the documentation and the quickstart guide this should be achievable with scoped injectors. So i tried to set it up in this way:

```
from sqlalchemy import create_engine, Connection

from pydio.injector import Injector
from pydio.provider import Provider


class ITodoItemStorage:
    pass


class SQLTodoStorage(ITodoItemStorage):

    def __init__(self, connection: Connection):
        self._conn = connection

    def create(self):
        raise ValueError()


provider = Provider()


@provider.provides('database', )
def make_database():
    return create_engine("postgresql://postgres:postrgres@localhost:5432/postgres")


@provider.provides('connection', scope="action")
def create_connection(injector):
    with injector.inject('database').begin() as c:
        try:
            yield c
        except:
            print("rollback")
            c.rollback()
        else:
            print("commit")
            c.commit()


@provider.provides(ITodoItemStorage, scope='action')
def make_sqlite_todo_storage(injector):
    connection = injector.inject('connection')
    return SQLTodoStorage(connection)


class Application:

    def __init__(self, injector):
        self._injector = injector

    def create(self):
        with self._injector.scoped('action') as inj:
            inj.inject(ITodoItemStorage).create()

if __name__ == '__main__':
    with Injector(provider) as injector:

        app = Application(injector)

        app.create()

```

But unfortunately, this is not working and the exception cannot be handled in the factory function.

Am I doing something completely wrong?

Thanks for the help in advance.

**Comment by mwiatrzyk** [comment]
Hello,

I'm glad you have found this lib useful :)

I will take a look at this issue as soon as possible. At first everything looks fine and most of examples in the docs are test cases by their own so I cannot tell right away what did not work here...

**Comment by mwiatrzyk** [comment]
Hello,

the issue is solved in v0.4.0.

BR,
Maciej
